### ISSUE TITLE
Type conversion fused with multiplication without extra copies

### ISSUE THREAD

**Issue by vadimkantorov** [numpy/numpy#17196]
I am looking to do int16<>float32 type conversions (frequent for audio sample conversions):
```
import numpy as np

x = np.array([1,2,3], dtype = 'int16')

smax = 32767
f2s_numpy = lambda signal, max = np.float32(smax): np.multiply(signal, smax).astype('int16')
s2f_numpy = lambda signal, max = np.float32(smax): np.divide(signal, max, dtype = 'float32')

f = s2f_numpy(x)
s = f2s_numpy(f)

print(x)
#[1 2 3]

print(s)
#[1 2 3]

print(f)
#[3.051851e-05 6.103702e-05 9.155553e-05]
```

I think `f2s_numpy` does a copy when converting to int16. As correctly suggested in https://stackoverflow.com/questions/63396575/fused-fast-conversion-from-int16-to-1-0-1-0-float32-range-in-numpy#comment112361047_63401786, `f2s_numpy = lambda signal: np.multiply(signal, np.float32(32767), dtype = 'int16')`  variant is not accepted: `numpy.core._exceptions.UFuncTypeError: Cannot cast ufunc 'multiply' input 0 from dtype('float32') to dtype('int16') with casting rule 'same_kind'`.

It would be good if this variant was also supported, since these conversions are quite frequent in practice.

**Comment by charris** [comment]
You can use the `casting='unsafe'` argument in the multiply function to override the default.
```
In [4]: np.multiply(x, np.float32(32767), dtype = 'int16', casting='unsafe')    
Out[4]: array([32767,    -2, 32765], dtype=int16)
```

**Comment by vadimkantorov** [comment]
Yep, but the result is incorrect :( `[32767,    -2, 32765]` instead of `[1, 2, 3]`

**Comment by seberg** [comment]
So I guess what you ask is for a different spelling of the `f2s` function?
```
def f2s_numpy(signal, scale=np.float32(2**16-1)):
     assert(signal.dtype == "float32")
     res = signal * scale
     return scale.astype(np.int16)
```
The `dtype` argument (as well as the more comprehensive, but less known signature argument) to ufuncs actually select the type the *operation is performed on* and not really the result dtype.

I think you can achieve what you want slightly more efficient with the following setup:
```
def f2s_numpy(signal, scale=np.float32(2**16-1)):
     assert(signal.dtype == "float32")
     res = np.empty_like(signal, dtype=np.int16)
     np.multiply(signal, scale, dtype=np.float32, out=res, casting="unsafe")
     return res
```
I am not sure how this will compare to a custom ufunc, which should be fairly straight forward in this case.

**Comment by seberg** [comment]
I don't think this is actionable, so closing.  Jit'ers like numba may be a path to a custom ufunc for the original poster.  Maybe they can even hook into NumPy directly one day.

But beyond optimizing the pattern that I gave above, I don't see what could be done in NumPy at this point.

**Comment by vadimkantorov** [comment]
This is maybe a question on controlling `compute_dtype` and `out_dtype` separately. I'd like that the passed dtype cast (f2s) is applied only before store and that the computation is still done in float32: (f * 32767.0)

Given that such casts are frequent (in audio/image processing / normalization), they ideally should happen without unneeded temp allocations/casts/memory passes.

Maybe this can be done by introducing a new, clear-named function: `linear_cast(x, out = ..., saturate : bool = False)`

**Comment by seberg** [comment]
> Also, numpy.astype does not support out= (for providing output buffer manually)

That functionality is covered by `copyto`.
